python/ray/dag/compiled_dag_node.py

```diff
@@ -517,25 +523,64 @@ def __init__(
             assert not isinstance(val, ChannelInterface)
 
         # Input reader to read input data from upstream DAG nodes.
-        self.input_reader: ReaderInterface = SynchronousReader(self.input_channels)
+        self.input_reader: Optional[ReaderInterface] = None
+        # NCCL P2P recv uses the NCCL channel instead of the input reader.
+        if not self.requires_nccl_read:
+            self.input_reader = SynchronousReader(self.input_channels)
```

Comment:
Instead of adding this extra logic here, it would be better if we can make sure `self.input_channels` is always empty for `self.requires_nccl_read` when reaching this line. Same for `self.output_channels` and `self.requires_nccl_write`.